src/EventListener/ExceptionListener.php

```diff
@@ -0,0 +1,307 @@
+                    // Throwing the page exception here is intentional as we have some error handler problems!
+                    // We should never end up here.
+                    // Fixing error handling code paths is more important than the original exception.
+                    throw $pageException;
+                }
+            }
+        }
+
+        return null;
+     }
+
+    /**
+     * Check if the exception is understood by us.
+     *
+     * @param \Exception $exception The exception to walk on.
+     *
+     * @return null|Response
+     */
+    private function checkHttpException($exception)
+    {
+        if (!$exception instanceof HttpExceptionInterface) {
+            return null;
+        }
+
+        $response = $this->tryToRenderContao404($exception);
+        if ($response) {
+            return $response;
+        }
+
+        // Determine if the class or any of the parents is known by us.
+        $candidates = array_intersect(
+            array_merge([get_class($exception)], class_parents($exception)),
+            array_keys(self::$exceptionTemplates)
+        );
+        if (!empty($candidates)) {
+            $template = self::$exceptionTemplates[array_shift($candidates)];
+            return $this->createTemplateResponseFromException($template, $exception);
+        }
+
+        // Unknown HttpExceptionInterface implementing exception, no way to handle.
+        return null;
+    }
+
+    /**
+     * Try to create a response for the given template.
+     *
+     * @param string                 $template  The name of the template to render.
+     * @param HttpExceptionInterface $exception The exception to render.
+     *
+     * @return Response
+     */
+    private function createTemplateResponseFromException($template, HttpExceptionInterface $exception)
+    {
+        return new Response(
+            $this->renderErrorTemplate($template),
+            $exception->getStatusCode(),
+            $exception->getHeaders()
+        );
+    }
+
+    /**
+     * Try to render an error template.
+     *
+     * @param string $template The template name. Will get searched at standard locations.
+     *
+     * @return string
+     */
+    private function renderErrorTemplate($template)
+    {
+        // TODO: make twig templates out of these.
+        if ($response = $this->tryReadTemplate(sprintf('%s/templates/%s.html5', $this->rootDir, $template))) {
+            return $response;
+        }
+
+        return $this->tryReadTemplate(
+            sprintf('%s/../Resources/contao/templates/backend/%s.html5', __DIR__, $template)
+        );
+    }
+
+    /**
+     * Try to read a template.
+     *
+     * @param string $template The file path to test.
+     *
+     * @return null|string
+     */
+    private function tryReadTemplate($template)
+    {
+        if (file_exists($template)) {
+            // Isolate the template parsing, the "unused" root dir variable will get used in the template.
```

Comment:
> The error templates should not derive the `$this` context of the exception listener.

This would probably be the better comment. Now I do understand :)